src/mongocxx/pool.hpp

```diff
@@ -109,6 +109,9 @@ class MONGOCXX_API pool {
     ///
     stdx::optional<entry> try_acquire();
 
+    // Attempts to create a new client pool using the uri. Throws an exception upon error.
+    void* construct_client_pool(const uri& uri);
```

Comment:
I was unable to get `mongoc_client_pool_t` to be the return type of this function.
```
/home/ubuntu/code/mongo-cxx-driver/src/mongocxx/../mongocxx/pool.hpp:112:5: error: 'mongoc_client_pool_t' does not name a type
  112 |     mongoc_client_pool_t* initialize_impl(const uri& uri);
```
It might be something small that I forgot to include, but that's the reason why this function returns a `void *` and then casts the return value in the initializer list above.

Reply:
> I was unable to get `mongoc_client_pool_t` to be the return type of this function.

The C driver API is explicitly hidden in public C++ driver headers. The C driver is intended to be an implementation detail in the driver. The private `.hh` headers and `.cpp` may use the C driver API.

Rather than make a private method on `pool.hpp`, I suggest making a function in `pool.cpp`:

```c++
static mongoc_client_pool_t* construct_client_pool(mongoc_uri_t* uri) {
    bson_error_t error;
    auto pool = libmongoc::client_pool_new_with_error(uri, &error);
    if (!pool) {
        // If constructing a client pool failed, throw an exception from the bson_error_t.
        throw_exception<operation_exception>(error);
    }

    return pool;
}
```